sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,765 @@
+buffer, which holds values of all specialization constants and should be used
+to retrieve their values in device code.
+
+This pointer is stored within `kernel_handler` object and it is initialized only
+if our target doesn't support native specialization constants.
+Since `kernel_handler` object is not captured by SYCL kernel funtion, it means
+that we are not able to employ some header-only solution here and need help of
+the compiler.
+
+DPC++ FE searches for functions marked with `sycl_kernel` attribute to handle
+them and turn into entry points of device code.
+
+
+#### DPC++ FE
+
+When we compile code for target which doesn't support native specialization
+constants, DPC++ FE should look for `kernel_handler` argument in functions
+marked as `sycl_kernel`. If such argument is present, it means that this kernel
+can access specialization constants and therefore we need to:
+- generate one more kernel argument for passing a buffer with specialization
+  constants values.
+- create `kernel_handler` object
+  
+  **TODO**: this item should be done for native specialization constants as
+  well, probably need to refactor the document to outline common parts into a
+  separate section.
+- initialize that `kernel_handler` object with newly created kernel argument
+- pass that `kernel_handler` object to user-provided SYCL kernel function
+
+So, having the following as the input:
+```
+template <typename KernelName, typename KernelType>
+__attribute__((sycl_kernel)) void
+kernel_single_task(const KernelType &KernelFunc, kernel_handler kh) {
+  KernelFunc(kh);
+}
+```
+DPC++ FE shoud tranform it into something like:
+
+```
+__kernel void KernelName(args_for_lambda_init, ..., char *specialization_constants_buffer) {
+  KernelType LocalLambdaClone = { args_for_lambda_init }; // We already do this
+  kernel_handler LocalKernelHandler;
+  LocalKernelHandler.__init_specialization_constants_buffer(specialization_constants_buffer);
+  // Re-used body of "sycl_kernel" function:
+  {
+     LocalLambdaClone(LocalKernelHandler);
+  }
+}
+```
+
+Besides that transformation, DPC++ FE should also provide information about that
+new kernel argument through integration header
+
+The new kernel argument `specialization_constants_buffer` should have
+corresponding entry in the `kernel_signatures` structure in the integration
+header. The param kind for this argument should be
+`kernel_param_kind_t:specialization_constants_buffer`.
+
+Example:
+```
+  const kernel_param_desc_t kernel_signatures[] = {
+   //--- _ZTSN2cl4sycl6detail19__pf_kernel_wrapperIZZ4mainENK3$_0clERNS0_7handlerEE6init_aEE
+   { kernel_param_kind_t::kind_std_layout, 8, 0 },
+   { kernel_param_kind_t::kind_accessor, 4062, 8 },
+   { kernel_param_kind_t::kind_specialization_constants_buffer, /*parameter_size_in_bytes= */ 8, /*offset_in_lambda=*/0},
+ };
+
+```
+
+Offset for this argument is zero since it has no any connected captured
```

Comment:
@romanovvlad If the offset is meaningless does it make sense to use an obviously meaningless number like unsigned-max? See https://github.com/intel/llvm/pull/3345#discussion_r592681484

Reply:
I think so.